Setting runOnDemandStartTimeout or runOnDemandCloseAfter for a path causes segfault/panic in v1.2.0

<!--
To increase the chance of your issue getting fixed, open an issue FOR EACH problem. Please do not report multiple problems in a single issue, otherwise they'll probably never get ALL fixed.

Please include all sections of this template into your issue, or it will be automatically closed.
-->

## Which version are you using?

v1.2.0

## Which operating system are you using?

<!-- fill checkboxes with a x. Example: [x] Linux -->

- [x] Linux amd64 standard
- [x] Linux amd64 Docker
- [ ] Linux arm64 standard
- [ ] Linux arm64 Docker
- [ ] Linux arm7 standard
- [ ] Linux arm7 Docker
- [ ] Linux arm6 standard
- [ ] Linux arm6 Docker
- [ ] Windows amd64 standard
- [ ] Windows amd64 Docker (WSL backend)
- [x] macOS amd64 standard
- [ ] macOS amd64 Docker
- [ ] Other (please describe)

## Describe the issue

Setting runOnDemandStartTimeout and/or runOnDemandCloseAfter for a specific path  causes a segfault/panic. This issue started with  v1.2.0 and was not an issue with  v1.1.1.

## Describe how to replicate the issue

<!--
the maintainers must be able to REPLICATE your issue to solve it - therefore, describe in a very detailed way how to replicate it.
-->


1. Set runOnDemandStartTimeout and/or runOnDemandCloseAfter for a specific path
2. start server

```
MTX_PATHS_MYPATH_RUNONDEMANDSTARTTIMEOUT=30s ./mediamtx 
```

## Did you attach the server logs?

<!--
Server logs are sometimes useful to identify the issue.
If you think this is the case, set the parameter 'logLevel' to 'debug' and attach the server logs.
-->

yes 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x16aeb7a]

goroutine 1 [running]:
github.com/bluenviron/mediamtx/internal/conf.(*StringDuration).UnmarshalJSON(0x0, {0xc0003748a0, 0x5, 0x8})
	/s/internal/conf/string_duration.go:28 +0x9a
github.com/bluenviron/mediamtx/internal/conf.(*StringDuration).UnmarshalEnv(0x19cc2a0?, {0xc000590158?, 0xc00016b7a0?}, {0xc000042089?, 0x0?})
	/s/internal/conf/string_duration.go:35 +0x5d
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0xc00013bd40?, {0xc00016b7a0, 0x28}, {0x19effe0?, 0xc0002d5380?, 0xc00003b998?})
	/s/internal/conf/env/env.go:33 +0x1202
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0xc00013bd40?, {0xc000374400, 0x10}, {0xc0000b3440?, 0xc0002d5180?, 0xc00058fd90?})
	/s/internal/conf/env/env.go:164 +0x1eb5
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0xc0000b3440?, {0xc000374400, 0x10}, {0xc00013bd40?, 0xc0002d5180?, 0x22d7f80?})
	/s/internal/conf/env/env.go:28 +0x12ef
github.com/bluenviron/mediamtx/internal/conf/env.Load({0xc000374400, 0x10}, {0xc0000b3440?, 0xc0002d5180})
	/s/internal/conf/env/env.go:224 +0x237
github.com/bluenviron/mediamtx/internal/conf.(*OptionalPath).UnmarshalEnv(0xc0002f2430?, {0xc000374400?, 0x10?}, {0x10?, 0x1978c5e?})
	/s/internal/conf/optional_path.go:64 +0x85
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x19fb660?, {0xc000374400, 0x10}, {0x19efde0?, 0xc0002f2430?, 0xc0005907f0?})
	/s/internal/conf/env/env.go:38 +0x13a4
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x19efde0?, {0xc000374400, 0x10}, {0x19fb660?, 0xc0002f2430?, 0xc00004206a?})
	/s/internal/conf/env/env.go:28 +0x12ef
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x19ca380?, {0xc0003743e5, 0x9}, {0xc0000b3300?, 0xc00015ef40?, 0x1067d39?})
	/s/internal/conf/env/env.go:146 +0x19dd
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x1af4880?, {0xc0003743e5, 0x9}, {0x19ca380?, 0xc00015ef40?, 0xc0003743e0?})
	/s/internal/conf/env/env.go:28 +0x12ef
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x1af4880?, {0x1af6d24, 0x3}, {0x1a99940?, 0xc00015e800?, 0x195ad00?})
	/s/internal/conf/env/env.go:164 +0x1eb5
github.com/bluenviron/mediamtx/internal/conf/env.loadEnvInternal(0x1a99940?, {0x1af6d24, 0x3}, {0x1af4880?, 0xc00015e800?, 0xc0001b6d18?})
	/s/internal/conf/env/env.go:28 +0x12ef
github.com/bluenviron/mediamtx/internal/conf/env.Load({0x1af6d24, 0x3}, {0x1a99940?, 0xc00015e800})
	/s/internal/conf/env/env.go:224 +0x237
github.com/bluenviron/mediamtx/internal/conf.Load({0x0, 0x0}, {0x22c5120, 0x5, 0x5})
	/s/internal/conf/conf.go:262 +0xb7
github.com/bluenviron/mediamtx/internal/core.New({0xc0000a01e0, 0x0, 0x0})
	/s/internal/core/core.go:138 +0x365
main.main()
	/s/main.go:11 +0x52
```

## Did you attach a network dump?

<!--
If the bug arises when using MediaMTX with an external hardware or software, the most helpful content you can provide is a dump of the data exchanged between the server and the target (network dump), that can be generated in this way:
1) Download wireshark (https://www.wireshark.org/)
2) Start capturing on the interface used for exchanging RTSP (if the server and the target software are both installed on your pc, the interface is probably "loopback", otherwise it's the one of your network card)
3) Start the server and replicate the issue
4) Stop capturing, save the result in .pcap format
5) Attach
-->

 no

--- a/internal/conf/env/env.go
+++ b/internal/conf/env/env.go
@@ -28,8 +28,14 @@ func loadEnvInternal(env map[string]string, prefix string, prv reflect.Value) er
 		return loadEnvInternal(env, prefix, prv.Addr())
 	}
 
+	rt := prv.Type().Elem()
+
 	if i, ok := prv.Interface().(Unmarshaler); ok {
 		if ev, ok := env[prefix]; ok {
+			if prv.IsNil() {
+				prv.Set(reflect.New(rt))
+				i = prv.Interface().(Unmarshaler)
+			}
 			err := i.UnmarshalEnv(prefix, ev)
 			if err != nil {
 				return fmt.Errorf("%s: %s", prefix, err)
@@ -42,8 +48,6 @@ func loadEnvInternal(env map[string]string, prefix string, prv reflect.Value) er
 		}
 		return nil
 	}
-
-	rt := prv.Type().Elem()
 
 	switch rt {
 	case reflect.TypeOf(""):
@@ -173,12 +177,12 @@ func loadEnvInternal(env map[string]string, prefix string, prv reflect.Value) er
 		switch {
 		case rt.Elem() == reflect.TypeOf(""):
 			if ev, ok := env[prefix]; ok {
-				if prv.IsNil() {
-					prv.Set(reflect.New(rt))
-				}
 				if ev == "" {
 					prv.Elem().Set(reflect.MakeSlice(prv.Elem().Type(), 0, 0))
 				} else {
+					if prv.IsNil() {
+						prv.Set(reflect.New(rt))
+					}
 					prv.Elem().Set(reflect.ValueOf(strings.Split(ev, ",")))
 				}
 			}
@@ -186,9 +190,6 @@ func loadEnvInternal(env map[string]string, prefix string, prv reflect.Value) er
 
 		case rt.Elem().Kind() == reflect.Struct:
 			if ev, ok := env[prefix]; ok && ev == "" { // special case: empty list
-				if prv.IsNil() {
-					prv.Set(reflect.New(rt))
-				}
 				prv.Elem().Set(reflect.MakeSlice(prv.Elem().Type(), 0, 0))
 			} else {
 				for i := 0; ; i++ {
@@ -213,13 +214,20 @@ func loadEnvInternal(env map[string]string, prefix string, prv reflect.Value) er
 	return fmt.Errorf("unsupported type: %v", rt)
 }
 
-// Load loads the configuration from the environment.
-func Load(prefix string, v interface{}) error {
+func loadWithEnv(env map[string]string, prefix string, v interface{}) error {
+	return loadEnvInternal(env, prefix, reflect.ValueOf(v).Elem())
+}
+
+func envToMap() map[string]string {
 	env := make(map[string]string)
 	for _, kv := range os.Environ() {
 		tmp := strings.SplitN(kv, "=", 2)
 		env[tmp[0]] = tmp[1]
 	}
+	return env
+}
 
-	return loadEnvInternal(env, prefix, reflect.ValueOf(v).Elem())
+// Load loads the configuration from the environment.
+func Load(prefix string, v interface{}) error {
+	return loadWithEnv(envToMap(), prefix, v)
 }

--- a/internal/conf/env/env_test.go
+++ b/internal/conf/env/env_test.go
@@ -9,13 +9,28 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-type subStruct struct {
-	MyParam int `json:"myParam"`
+func stringPtr(v string) *string {
+	return &v
 }
 
-type mapEntry struct {
-	MyValue  string    `json:"myValue"`
-	MyStruct subStruct `json:"myStruct"`
+func intPtr(v int) *int {
+	return &v
+}
+
+func uint64Ptr(v uint64) *uint64 {
+	return &v
+}
+
+func boolPtr(v bool) *bool {
+	return &v
+}
+
+func float64Ptr(v float64) *float64 {
+	return &v
+}
+
+func durationPtr(v time.Duration) *time.Duration {
+	return &v
 }
 
 type myDuration time.Duration
@@ -40,106 +55,150 @@ func (d *myDuration) UnmarshalEnv(_ string, v string) error {
 	return d.UnmarshalJSON([]byte(`"` + v + `"`))
 }
 
+type subStruct struct {
+	MyParam int `json:"myParam"`
+}
+
+type mapEntry struct {
+	MyValue  string    `json:"myValue"`
+	MyStruct subStruct `json:"myStruct"`
+}
+
 type mySubStruct struct {
 	URL      string `json:"url"`
 	Username string `json:"username"`
 	Password string `json:"password"`
+	MyInt2   int    `json:"myInt2"`
 }
 
 type testStruct struct {
-	MyString              string               `json:"myString"`
-	MyInt                 int                  `json:"myInt"`
-	MyFloat               float64              `json:"myFloat"`
-	MyBool                bool                 `json:"myBool"`
-	MyDuration            myDuration           `json:"myDuration"`
-	MyMap                 map[string]*mapEntry `json:"myMap"`
-	MySlice               []string             `json:"mySlice"`
-	MySliceEmpty          []string             `json:"mySliceEmpty"`
-	MySliceSubStruct      []mySubStruct        `json:"mySliceSubStruct"`
-	MySliceSubStructEmpty []mySubStruct        `json:"mySliceSubStructEmpty"`
+	MyString                 string               `json:"myString"`
+	MyStringOpt              *string              `json:"myStringOpt"`
+	MyInt                    int                  `json:"myInt"`
+	MyIntOpt                 *int                 `json:"myIntOpt"`
+	MyUint                   uint64               `json:"myUint"`
+	MyUintOpt                *uint64              `json:"myUintOpt"`
+	MyFloat                  float64              `json:"myFloat"`
+	MyFloatOpt               *float64             `json:"myFloatOpt"`
+	MyBool                   bool                 `json:"myBool"`
+	MyBoolOpt                *bool                `json:"myBoolOpt"`
+	MyDuration               myDuration           `json:"myDuration"`
+	MyDurationOpt            *myDuration          `json:"myDurationOpt"`
+	MyDurationOptUnset       *myDuration          `json:"myDurationOptUnset"`
+	MyMap                    map[string]*mapEntry `json:"myMap"`
+	MySliceString            []string             `json:"mySliceString"`
+	MySliceStringEmpty       []string             `json:"mySliceStringEmpty"`
+	MySliceStringOpt         *[]string            `json:"mySliceStringOpt"`
+	MySliceStringOptUnset    *[]string            `json:"mySliceStringOptUnset"`
+	MySliceSubStruct         []mySubStruct        `json:"mySliceSubStruct"`
+	MySliceSubStructEmpty    []mySubStruct        `json:"mySliceSubStructEmpty"`
+	MySliceSubStructOpt      *[]mySubStruct       `json:"mySliceSubStructOpt"`
+	MySliceSubStructOptUnset *[]mySubStruct       `json:"mySliceSubStructOptUnset"`
+	Unset                    *bool                `json:"unset"`
 }
 
 func TestLoad(t *testing.T) {
-	os.Setenv("MYPREFIX_MYSTRING", "testcontent")
-	defer os.Unsetenv("MYPREFIX_MYSTRING")
+	env := map[string]string{
+		"MYPREFIX_MYSTRING":                       "testcontent",
+		"MYPREFIX_MYSTRINGOPT":                    "testcontent2",
+		"MYPREFIX_MYINT":                          "123",
+		"MYPREFIX_MYINTOPT":                       "456",
+		"MYPREFIX_MYUINT":                         "8910",
+		"MYPREFIX_MYUINTOPT":                      "112313",
+		"MYPREFIX_MYFLOAT":                        "15.2",
+		"MYPREFIX_MYFLOATOPT":                     "16.2",
+		"MYPREFIX_MYBOOL":                         "yes",
+		"MYPREFIX_MYBOOLOPT":                      "false",
+		"MYPREFIX_MYDURATION":                     "22s",
+		"MYPREFIX_MYDURATIONOPT":                  "30s",
+		"MYPREFIX_MYMAP_MYKEY":                    "",
+		"MYPREFIX_MYMAP_MYKEY2_MYVALUE":           "asd",
+		"MYPREFIX_MYMAP_MYKEY2_MYSTRUCT_MYPARAM":  "456",
+		"MYPREFIX_MYSLICESTRING":                  "val1,val2",
+		"MYPREFIX_MYSLICESTRINGEMPTY":             "",
+		"MYPREFIX_MYSLICESTRINGOPT":               "aa",
+		"MYPREFIX_MYSLICESUBSTRUCT_0_URL":         "url1",
+		"MYPREFIX_MYSLICESUBSTRUCT_0_USERNAME":    "user1",
+		"MYPREFIX_MYSLICESUBSTRUCT_0_PASSWORD":    "pass1",
+		"MYPREFIX_MYSLICESUBSTRUCT_1_URL":         "url2",
+		"MYPREFIX_MYSLICESUBSTRUCT_1_PASSWORD":    "pass2",
+		"MYPREFIX_MYSLICESUBSTRUCTEMPTY":          "",
+		"MYPREFIX_MYSLICESUBSTRUCTOPT_1_PASSWORD": "pwd",
+	}
 
-	os.Setenv("MYPREFIX_MYINT", "123")
-	defer os.Unsetenv("MYPREFIX_MYINT")
-
-	os.Setenv("MYPREFIX_MYFLOAT", "15.2")
-	defer os.Unsetenv("MYPREFIX_MYFLOAT")
-
-	os.Setenv("MYPREFIX_MYBOOL", "yes")
-	defer os.Unsetenv("MYPREFIX_MYBOOL")
-
-	os.Setenv("MYPREFIX_MYDURATION", "22s")
-	defer os.Unsetenv("MYPREFIX_MYDURATION")
-
-	os.Setenv("MYPREFIX_MYMAP_MYKEY", "")
-	defer os.Unsetenv("MYPREFIX_MYMAP_MYKEY")
-
-	os.Setenv("MYPREFIX_MYMAP_MYKEY2_MYVALUE", "asd")
-	defer os.Unsetenv("MYPREFIX_MYMAP_MYKEY2_MYVALUE")
-
-	os.Setenv("MYPREFIX_MYMAP_MYKEY2_MYSTRUCT_MYPARAM", "456")
-	defer os.Unsetenv("MYPREFIX_MYMAP_MYKEY2_MYSTRUCT_MYPARAM")
-
-	os.Setenv("MYPREFIX_MYSLICE", "val1,val2")
-	defer os.Unsetenv("MYPREFIX_MYSLICE")
-
-	os.Setenv("MYPREFIX_MYSLICEEMPTY", "")
-	defer os.Unsetenv("MYPREFIX_MYSLICEEMPTY")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCT_0_URL", "url1")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCT_0_URL")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCT_0_USERNAME", "user1")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCT_0_USERNAME")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCT_0_PASSWORD", "pass1")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCT_0_PASSWORD")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCT_1_URL", "url2")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCT_1_URL")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCT_1_PASSWORD", "pass2")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCT_1_PASSWORD")
-
-	os.Setenv("MYPREFIX_MYSLICESUBSTRUCTEMPTY", "")
-	defer os.Unsetenv("MYPREFIX_MYSLICESUBSTRUCTEMPTY")
+	for key, val := range env {
+		os.Setenv(key, val)
+		defer os.Unsetenv(key)
+	}
 
 	var s testStruct
 	err := Load("MYPREFIX", &s)
 	require.NoError(t, err)
 
-	require.Equal(t, "testcontent", s.MyString)
-	require.Equal(t, 123, s.MyInt)
-	require.Equal(t, 15.2, s.MyFloat)
-	require.Equal(t, true, s.MyBool)
-	require.Equal(t, 22*myDuration(time.Second), s.MyDuration)
-
-	_, ok := s.MyMap["mykey"]
-	require.Equal(t, true, ok)
-
-	v, ok := s.MyMap["mykey2"]
-	require.Equal(t, true, ok)
-	require.Equal(t, "asd", v.MyValue)
-	require.Equal(t, 456, v.MyStruct.MyParam)
-
-	require.Equal(t, []string{"val1", "val2"}, s.MySlice)
-	require.Equal(t, []string{}, s.MySliceEmpty)
-
-	require.Equal(t, []mySubStruct{
-		{
-			URL:      "url1",
-			Username: "user1",
-			Password: "pass1",
+	require.Equal(t, testStruct{
+		MyString:      "testcontent",
+		MyStringOpt:   stringPtr("testcontent2"),
+		MyInt:         123,
+		MyIntOpt:      intPtr(456),
+		MyUint:        8910,
+		MyUintOpt:     uint64Ptr(112313),
+		MyFloat:       15.2,
+		MyFloatOpt:    float64Ptr(16.2),
+		MyBool:        true,
+		MyBoolOpt:     boolPtr(false),
+		MyDuration:    22000000000,
+		MyDurationOpt: (*myDuration)(durationPtr(30000000000)),
+		MyMap: map[string]*mapEntry{
+			"mykey": {
+				MyValue: "",
+				MyStruct: subStruct{
+					MyParam: 0,
+				},
+			},
+			"mykey2": {
+				MyValue: "asd",
+				MyStruct: subStruct{
+					MyParam: 456,
+				},
+			},
 		},
-		{
-			URL:      "url2",
-			Password: "pass2",
+		MySliceString: []string{
+			"val1",
+			"val2",
 		},
-	}, s.MySliceSubStruct)
+		MySliceStringEmpty: []string{},
+		MySliceStringOpt:   &[]string{"aa"},
+		MySliceSubStruct: []mySubStruct{
+			{
+				URL:      "url1",
+				Username: "user1",
+				Password: "pass1",
+			},
+			{
+				URL:      "url2",
+				Username: "",
+				Password: "pass2",
+			},
+		},
+		MySliceSubStructEmpty: []mySubStruct{},
+	}, s)
+}
 
-	require.Equal(t, []mySubStruct{}, s.MySliceSubStructEmpty)
+func FuzzLoad(f *testing.F) {
+	f.Add("MYPREFIX_MYINT", "a")
+	f.Add("MYPREFIX_MYUINT", "a")
+	f.Add("MYPREFIX_MYFLOAT", "a")
+	f.Add("MYPREFIX_MYBOOL", "a")
+	f.Add("MYPREFIX_MYSLICESUBSTRUCT_0_MYINT2", "a")
+	f.Add("MYPREFIX_MYDURATION", "a")
+	f.Add("MYPREFIX_MYDURATION_A", "a")
+
+	f.Fuzz(func(t *testing.T, key string, val string) {
+		env := map[string]string{
+			key: val,
+		}
+
+		var s testStruct
+		loadWithEnv(env, "MYPREFIX", &s) //nolint:errcheck
+	})
 }

